### PR TITLE
HBASE-27053 IOException during caching of uncompressed block to the b…

### PR DESCRIPTION
**Pull request by ss77892**
…lock cache.

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 53s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  4s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ branch-2 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 50s |  branch-2 passed  |
| +1 :green_heart: |  compile  |   0m 14s |  branch-2 passed  |
| +1 :green_heart: |  shadedjars  |   3m 53s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 15s |  branch-2 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m  7s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 13s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 13s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   3m 54s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 12s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 22s |  hbase-common in the patch passed.  |
|  |   |  17m 44s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4453/1/artifact/yetus-jdk8-hadoop2-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4453 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux f4525da6e78b 5.4.0-1025-aws #25~18.04.1-Ubuntu SMP Fri Sep 11 12:03:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | branch-2 / 8a3a94205b |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4453/1/testReport/ |
| Max. process+thread count | 246 (vs. ulimit of 12500) |
| modules | C: hbase-common U: hbase-common |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4453/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 11s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  4s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ branch-2 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 49s |  branch-2 passed  |
| +1 :green_heart: |  compile  |   0m 18s |  branch-2 passed  |
| +1 :green_heart: |  shadedjars  |   3m 52s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 18s |  branch-2 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 26s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 18s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 18s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   3m 51s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 16s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 51s |  hbase-common in the patch passed.  |
|  |   |  18m  4s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4453/1/artifact/yetus-jdk11-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4453 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux 555d363302f7 5.4.0-90-generic #101-Ubuntu SMP Fri Oct 15 20:00:55 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | branch-2 / 8a3a94205b |
| Default Java | AdoptOpenJDK-11.0.10+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4453/1/testReport/ |
| Max. process+thread count | 231 (vs. ulimit of 12500) |
| modules | C: hbase-common U: hbase-common |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4453/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 53s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ branch-2 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 37s |  branch-2 passed  |
| +1 :green_heart: |  compile  |   0m 30s |  branch-2 passed  |
| +1 :green_heart: |  checkstyle  |   0m 15s |  branch-2 passed  |
| +1 :green_heart: |  spotless  |   0m 46s |  branch has no errors when running spotless:check.  |
| +1 :green_heart: |  spotbugs  |   0m 37s |  branch-2 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 57s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 28s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 28s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 11s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |   8m  4s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.1.  |
| -1 :x: |  spotless  |   0m 11s |  patch has 22 errors when running spotless:check, run spotless:apply to fix.  |
| +1 :green_heart: |  spotbugs  |   0m 31s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 14s |  The patch does not generate ASF License warnings.  |
|  |   |  23m  6s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4453/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/4453 |
| Optional Tests | dupname asflicense javac spotbugs hadoopcheck hbaseanti spotless checkstyle compile |
| uname | Linux 7ed89deb9667 5.4.0-1071-aws #76~18.04.1-Ubuntu SMP Mon Mar 28 17:49:57 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | branch-2 / 8a3a94205b |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
| spotless | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4453/1/artifact/yetus-general-check/output/patch-spotless.txt |
| Max. process+thread count | 60 (vs. ulimit of 12500) |
| modules | C: hbase-common U: hbase-common |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-4453/1/console |
| versions | git=2.17.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-common/src/main/java/org/apache/hadoop/hbase/io/encoding/HFileBlockDefaultDecodingContext.java [hunk 1]
```diff
@@ -103,6 +103,11 @@ public void prepareDecoding(int onDiskSizeWithoutHeader, int uncompressedSizeWit
             compression.createDecompressionStream(dataInputStream, decompressor, 0)) {
             BlockIOUtils.readFullyWithHeapBuffer(is, blockBufferWithoutHeader,
               uncompressedSizeWithoutHeader);
+            // The buffer has allocated space for checksum, but it's not always filled. That might
```
Comment (joshelser) [inline hbase-common/src/main/java/org/apache/hadoop/hbase/io/encoding/HFileBlockDefaultDecodingContext.java:106]: Yeah, I see this is getting allocated in HFileBlock#allocateBuffer. The thing I'm left wondering if why `BlockIOUtils.readFullyWithHeapBuffer(..)` is only _sometimes_ copying the checksum bytes after the uncompressed datablock. I guess that is the outstanding question :)

**Review by joshelser — COMMENTED** [review]
I know you told me offline that you weren't able to easily reproduce this (end to end) in a unit test. What about adding a contrived unit test which can validate that the filling of the bytebuff when the data is compressed works as expected?

@Apache9 you have any thoughts here (or know another dev who might have some thoughts)?

In general, I'm inclined to say this is "good", but I know that's based on internal work that Sergey has done here.

File: hbase-common/src/main/java/org/apache/hadoop/hbase/io/encoding/HFileBlockDefaultDecodingContext.java (hunk 1 above)
Comment (bbeaudreault) [reply hbase-common/src/main/java/org/apache/hadoop/hbase/io/encoding/HFileBlockDefaultDecodingContext.java:106]: I agree we should understand that.

**Comment by bbeaudreault** [comment]
@joshelser @Apache9 @ss77892 I decided to take a look at this, as we're trying to push forward on BucketCache in prod and this was an issue we see every time a region is split. I believe I have the real root cause and suggested fix, detailed below:

First question, why are we even getting into this point where a read does not get a cache hit, but then sees a conflict when caching it's result from disk? This is covered by the javadoc in `BlockCacheUtils#shouldReplaceExistingCacheBlock`:

>   * Because of the region splitting, it's possible that the split key locate in the middle of a
>   * block. So it's possible that both the daughter regions load the same block from their parent
>   * HFile. When pread, we don't force the read to read all of the next block header. So when two
>   * threads try to cache the same block, it's possible that one thread read all of the next block
>   * header but the other one didn't. if the already cached block hasn't next block header but the
>   * new block to cache has, then we can replace the existing block with the new block for better
>   * performance.(HBASE-20447)

Next question, why are we supposedly only _sometimes_ not reading the checksum? In actuality, we _never_ are reading the checksum, but the buffer always includes space for them. The unpacked HFileBlock's buffer always has random data at the end since we're allocating space for the checksum, but not filling it. This only becomes a problem in the split case because otherwise we never tried to double cache a block and we never reference the checksum once cached.  See below analysis:

1. The checksums are used when reading from disk, in `HFileBlock#readBlockDataInternal`. Beyond that, I don't see much use of checksums. 
2. In `HFileBlock#unpack`, we allocate a buffer for the decompression, with capacity `headerSize + uncompressedSizeWithoutHeader + cksumBytes`. Before passing to decoder, we put the header onto the buffer.
3. Next, we pass `getUncompressedSizeWithoutHeader()` to `HFileBlockDecodingContext#prepareEncoding`. The javadoc for this function says "the uncompressed size of data part (header and **checksum excluded**).". This value is used in Compression.decompress, to determine how many bytes to readFully into our destination buffer. 
4. At this point we immediately see a discrepancy -- the allocated buffer includes space for the checksum, but we are only writing headerSize and uncompressedSizeWithoutHeader into that buffer.

Let's validate that analysis:

1. On the write side, take a look at `HFileBlock#finishBlockAndWriteHeaderAndData`, the `onDiskBlockBytesWithHeader` is written out first, and then separately the checksums are written out.
2. The `onDiskBlockBytesWithHeader` is the blob which includes header and compressed bytes.
3. On reading, the `uncompressedSizeWithoutHeader` is read from that header using`buf.getInt(Header.UNCOMPRESSED_SIZE_WITHOUT_HEADER_INDEX)` (index 12). 
4. Looking back at the writing side, `HFileBlock#finishBlock` passes `baosInMemory.size()` as uncompressedSize to `putHeader`. `baosInMemory` does not include the checksumming, which is calculated _after_ writing the header in `finishBlock`. So the value read in step 3 above does not include checksum length.
5. So at this point I'm confident that the javadoc mentioned above is accurate, uncompressedSizeWithoutHeader does not include checksum.
6. However, as mentioned in the original analysis, we're allocating a buffer that includes `cksumBytes`, and we're passing into the encoder `unpacked.getBufferWithoutHeader(true)`, which includes the checksum part of that buffer.

Ok so at this point we can see that Compression.decompress is being passed an `uncompressedSize` which doesn't include checksum, and a buffer which includes space for a checksum. The `uncompressedSize` dictates how many bytes are read into that buffer, which means the checksum section at the end will be left untouched.

If BucketCache is enabled, all DATA block reads go to ByteBuffAllocator, which re-uses buffers of capacity 65k. When a buffer is pulled from the pool, `clear()` is called, which doesnt actually zero out the bytes. It just resets the pointers.

So if you allocate a buffer and don't fill the entire requested capacity, you will have junk left over. It's important to allocate only what you need, and always fill what you allocate. Otherwise we hit the problem we're facing in unpack.

The fix here is simple -- change HFileBlock#allocateBuffer to not include `cksumBytes`.  This will also save cpu cycles over the current fix in this PR, because we don't need to unnecessarily zero out the extra bytes.

I verified all of this with a new test method in the `TestHFileBlockUnpack` that I recently added for HBASE-27170. All you need to do is write a block, and then read that same block twice and expect the underlying buffer to be identical. See below:

```java
 @Test
  public void itUnpacksIdenticallyEachTime() throws IOException {
    Path path = new Path(TEST_UTIL.getDataTestDir(), name.getMethodName());
    int totalSize = createTestBlock(path);

    // Allocate a bunch of random buffers, so we can be sure that unpack will only have "dirty"
    // buffers to choose from when allocating itself.
    Random random = new Random();
    byte[] temp = new byte[HConstants.DEFAULT_BLOCKSIZE];
    List<ByteBuff> buffs = new ArrayList<>();
    for (int i = 0; i < 10; i++) {
      ByteBuff buff = allocator.allocate(HConstants.DEFAULT_BLOCKSIZE);
      random.nextBytes(temp);
      buff.put(temp);
      buffs.add(buff);
    }

    buffs.forEach(ByteBuff::release);

    // read the same block twice. we should expect the underlying buffer below to
    // be identical each time
    HFileBlock blockOne = readBlock(path, totalSize);
    HFileBlock blockTwo = readBlock(path, totalSize);

    ByteBuff bufferOne = blockOne.getBufferWithoutHeader(true);
    ByteBuff bufferTwo = blockTwo.getBufferWithoutHeader(true);

    // This assertion should succeed, but it fails on master. It will succeed once
    // cksumBytes is removed from HFileBlock#allocateBuffer.
    assertEquals(0, ByteBuff.compareTo(bufferOne, 0, bufferOne.limit(), bufferTwo, 0 ,bufferTwo.limit()));
  }
```

**Comment by bbeaudreault** [comment]
@ss77892 do you want to give the above a try, or would you like me to take this over? Happy to do either, but would love to get this fixed.

**Comment by huaxiangsun** [comment]
@bbeaudreault Nice analysis! Could you post the analysis at the jira so it wont get lost?

**Comment by bbeaudreault** [comment]
Done @huaxiangsun, thanks for the suggestion.

**Pull request closed without merging**